crates/uv/src/commands/tool/upgrade.rs

```diff
@@ -101,11 +136,26 @@ pub(crate) async fn upgrade(
         writeln!(printer.stderr(), "Nothing to upgrade")?;
     }
 
+    if let (true, Some(python)) = (did_upgrade, python) {
+        writeln!(
+            printer.stderr(),
+            "Upgraded build environment for {} to Python {}",
+            if names.len() == 1 {
+                format!("{}", names.first().unwrap().bold())
+            } else {
+                "all tools".bold().to_string()
+            },
+            python
+        )?;
+    }
+
     Ok(ExitStatus::Success)
 }
 
 async fn upgrade_tool(
     name: &PackageName,
+    interpreter: &Option<Interpreter>,
+    python_request: &Option<PythonRequest>,
```

Comment:
In general, prefer `Option<&T>` over `&Option<T>`. The latter can always be converted to the former, but the opposite isn't true. I.e., if you have `&Option<PythonRequest>`, you can cast to `Option<&PythonRequest>` via `as_ref()`. But if you have `Option<&PythonRequest>`, you can't get an `&Option<PythonRequest>`.